### ISSUE TITLE
Support for multiple notations with the same key

### ISSUE THREAD

**Issue by wiktor-k** [openpgpjs/openpgpjs#897]
Hello,

Currently it's impossible to access multiple [notations](https://tools.ietf.org/html/rfc4880#section-5.2.3.16) if they have the same key as the last one always overwrites all previous values.

I've added multiple signature notations with `gpg` and they are displayed well in CLI and in packet dump:

```
$ gpg --list-packets test.gpg 
# off=0 ctb=99 tag=6 hlen=3 plen=269
:public key packet:
	version 4, algo 1, created 1557150010, expires 0
	pkey[0]: [2048 bits]
	pkey[1]: [17 bits]
	keyid: 579D01B3ABE1540E
# off=272 ctb=b4 tag=13 hlen=2 plen=16
:user ID packet: "Testt <test@exa>"
# off=290 ctb=89 tag=2 hlen=3 plen=394
:signature packet: algo 1, keyid 579D01B3ABE1540E
(...truncated...)
	hashed subpkt 33 len 21 (issuer fpr v4 198C722A4BAC336E9DAAAE44579D01B3ABE1540E)
	hashed subpkt 2 len 4 (sig created 2019-05-06)
	hashed subpkt 20 len 25 (notation: test@example.com=2) <---- here
	hashed subpkt 20 len 25 (notation: test@example.com=3) <---- and here
	subpkt 16 len 8 (issuer key ID 579D01B3ABE1540E)
	data: [2048 bits]
```

Reading the same key with OpenPGP.js with this snippet of code:

```js
openpgp.key.read(fs.readFileSync('test.gpg')).then(k => k.keys[0].users[0].selfCertifications).then(console.log.bind(console), console.error.bind(console));
```

Prints the following info:

```
[ Signature {
    tag: 2,
    version: 4,
(...truncated...)
    issuerKeyId: Keyid { bytes: 'W },
    notation: { 'test@example.com': '3' }, <--- here
    preferredHashAlgorithms: [ 10, 9, 8, 11, 2 ],
    preferredCompressionAlgorithms: [ 2, 3, 1 ],
```

As you can see only the last notation value for that key was preserved (`test@example.com=3`) while the second one (`test@example.com=2`) is nowhere to be found.

[RFC 4880](https://tools.ietf.org/html/rfc4880#section-5.2.3.16) doesn't have any guidance as of what to do when there are multiple notations with the same key.

I see three possible solutions to this problem:
  - convert `notation` to store array of values instead of single one, disadvantage: this breaks public API,
  - convert `notation` to store either single value, if there is only one notation of that key, or an array of values if there are multiple ones, this still breaks API in rare cases that there are multiple notations,
  - expose additional key, e.g. `notations` that always uses arrays.
  - something else?

I volunteer to implement the fix if one of these solutions is acceptable.

**Comment by twiss** [comment]
I wouldn't be terribly opposed to replacing `notation` with `notations = [['test@example.com', '2'], ['test@example.com', '3']]`. The advantage of that is that if you don't care about the multiple values, you can do `new Map(notations).get('test@example.com') // -> '3'`.

Out of curiosity though, do you have a use-case for this, and/or do you feel like the notation subpacket is meant to be used like this? If you're defining a new property, you could e.g. define it as a comma-separated list, or a JSON value.

It's true that GPG prints both of them, but it's essentially just logging what it encounters, not really providing an API to access the data.

--- 

Btw, looking at the code, there's some other stuff we ought to fix - e.g. we shouldn't be checking whether `bytes[mypos] === 0x80` but instead whether `bytes[mypos] & 0x80`. And if that flag isn't set we should just return a Uint8Array. (We could also provide two APIs, one which is always `[['name', Uint8Array(value)], ...]` and the current one as-is.) Lmk if you have strong opinions on this.

Also we should be throwing if `critical` is set and we don't know know the notation name. (There's a list at https://github.com/gpg/gnupg/blob/f43560a4d6cd44824654989ae043b346910e7a43/g10/parse-packet.c#L199.)

Obviously don't feel obligated to fix all of the above too :) but if you feel like it, feel welcome.

**Comment by wiktor-k** [comment]
> Out of curiosity though, do you have a use-case for this, and/or do you feel like the notation subpacket is meant to be used like this? If you're defining a new property, you could e.g. define it as a comma-separated list, or a JSON value.

Actually all PRs I filed to OpenPGP.js is because I had some issue with an edge case I wanted to use, it's not esoteric :)

If you're interested I'm storing "social proof" links (like Keybase, but distributed) in these notations, such as:

```
proof@metacode.biz=https://gist.github.com/wiktor-k/389d589dd19250e1f9a42bc3d5d40c16
proof@metacode.biz=https://news.ycombinator.com/user?id=fasm
proof@metacode.biz=https://www.reddit.com/r/KeybaseProofs/comments/2d28r8/my_keybase_proof_redditfaragon_keybasefaragon/#Faragon
proof@metacode.biz=https://mastodon.social/@wiktor
proof@metacode.biz=dns:metacode.biz?type=TXT
```

And I'd like people to be able to work with them individually, that is, remove one of them without touching others (it's possible in `gpg --edit-key ...`). They also look quite readable, defining ad-hoc format wouldn't look that nice. Using JSON is a good idea but it's would look a little bit bad and I don't want to push JSON dependency to clients.

> It's true that GPG prints both of them, but it's essentially just logging what it encounters, not really providing an API to access the data.

Good point w.r.t. `list-packets` but `gpg --edit-key` has a fully functioning command-line interface to manipulate that (including adding, removing...).

> I wouldn't be terribly opposed to replacing notation with notations = [['test@example.com', '2'], ['test@example.com', '3']]. The advantage of that is that if you don't care about the multiple values, you can do new Map(notations).get('test@example.com') // -> '3'.

If you don't mind I will follow-up with a PR implementing that proposal (if no-one else objects).

> Btw, looking at the code, there's some other stuff we ought to fix

Yep, I can definitely fix this stuff up while I'm touching notation code.

**Comment by twiss** [comment]
(Fixed by #901)
